File: sample/src/main/java/com/sunmi/samples/printerx/MainActivity.java
package com.sunmi.samples.printerx;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;

import com.sunmi.printerx.PrinterSdk;
import com.sunmi.printerx.SdkException;

import java.util.List;

public class MainActivity extends AppCompatActivity {

    /**
     * 指定操作的商米打印机设备
     */
    public PrinterSdk.Printer selectPrinter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_printer);
        if(savedInstanceState == null) {
            switchFragment(MainFragment.class);
        }

        initPrinter();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        PrinterSdk.getInstance().destroy();
    }

    /**
     * 初始化默认打印机作为待操作的打印设备
     */
    private void initPrinter() {
        try {
            PrinterSdk.getInstance().getPrinter(this, new PrinterSdk.PrinterListen() {
                @Override
                public void onDefPrinter(PrinterSdk.Printer printer) {
                    selectPrinter = printer;
                }

                @Override
                public void onPrinters(List<PrinterSdk.Printer> printers) {

                }
            });
        } catch (SdkException e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.log_swtich, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_log_off:
                PrinterSdk.getInstance().log(false, null);
                break;
            case R.id.action_log_on:
                PrinterSdk.getInstance().log(true, "Test");
                break;
            default:
                return super.onOptionsItemSelected(item);
        }
        return true;
    }

    public void switchFragment(Class<? extends Fragment> fragmentClass) {
        getSupportFragmentManager().beginTransaction()
                .setReorderingAllowed(true)
                .replace(R.id.container, fragmentClass, null)
                .commit();
    }

    public void switchFragment(Class<? extends Fragment> fragmentClass, boolean isBack) {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction()
                .setReorderingAllowed(true)
                .replace(R.id.container, fragmentClass, null);
        if(isBack) {
            transaction.addToBackStack(fragmentClass.getSimpleName());
        }
        transaction.commit();
    }
}

File: sample-ktx/src/main/java/com/sunmi/samples/printerx/MainActivity.kt
package com.sunmi.samples.printerx

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import android.view.Menu
import android.view.MenuItem
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.Lifecycle
import com.sunmi.printerx.PrinterSdk
import com.sunmi.samples.printerx.ui.main.MainFragment

class MainActivity : AppCompatActivity() {

    private val handler by lazy {
        Handler(mainLooper)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_printer)
        if (savedInstanceState == null) {
            switchFragment(MainFragment.newInstance(), false)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.log_swtich, menu)
        return super.onCreateOptionsMenu(menu)
    }

    /**
     * 调试期间可通过开启log检查接口调用不打印或失败的原因
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when(item.itemId) {
            R.id.action_log_off-> PrinterSdk.getInstance().log(false, null)
            R.id.action_log_on-> PrinterSdk.getInstance().log(true, null)
            else -> return super.onOptionsItemSelected(item)
        }
        return true
    }

    fun switchFragment(
        fragment: Fragment,
        addToBackStack: Boolean = true,
        clearStack: Boolean = true,
    ) {
        handler.post {
            if (clearStack) {
                if (lifecycle.currentState != Lifecycle.State.RESUMED) {
                    fixBug()
                }
                var r = supportFragmentManager.popBackStackImmediate(
                    null,
                    FragmentManager.POP_BACK_STACK_INCLUSIVE
                )
            }
            var transaction = supportFragmentManager.beginTransaction()
            transaction.replace(R.id.container, fragment, fragment::class.java.name)
            if (addToBackStack) {
                transaction.addToBackStack(fragment::class.java.name)
            }
            transaction.commitAllowingStateLoss()
            supportFragmentManager.executePendingTransactions()
        }

    }

    fun fixBug() {
        try {
            var aClass = FragmentManager::class.java
            var method = aClass.getDeclaredMethod("noteStateNotSaved")
            method.setAccessible(true)
            method.invoke(supportFragmentManager)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}